kep/624-disk-io-aware-scheduling/README.md

```diff
@@ -0,0 +1,257 @@
+```
+type Normalizer interface {
+	Name() string
+	EstimateRequest(ioRequest string) string
+}
+```
+Here is a sample implementation.  
+``` go
+type normalizer struct {}
+type IORequest struct {
+  rbps string
+  wbps string
+  blockSize string
+}
+
+func (n normalizer) Name() string {
+	return "Intel P4510 NVMe Disk"
+}
+
+// ioRequest example: {"rbps": "30M", "wbps": "20M", "blocksize": "4k"}
+func (n normalizer) EstimateRequest(ioRequest) string {
+  var req  = &IORequest{}
+  _ = json.Unmarshal([]byte(ioRequest), req)
+  resp, _ := n.normalize(req)
+  normalized, _ := json.Marshal(resp)
+  return normalized
+}
+
+// customized normalization method
+func (n normalizer) normalize(ioRequest *IORequest) (*IORequest, error) {
+  return &IORequest{
+    rbps: ioRequest.rbps * coefficientA,
+    wbps: ioRequest.wbps * coefficientB,
+  }, nil
+}
+```
+
+ The IO Driver updates each node disk’s normalized available IO capacity to the API server and the scheduler plugin watches the info through the `NodeDiskIOInfo` CR and stores it in the plugin’s local cache. 
+The disk IO scheduler plugin consists of the following parts.
+
+#### Filter Plugin
+
+During the filter phase, the scheduler plugin passes the PodSpec with a disk IO BW request (as shown below) to the corresponding normalization function and gets back a normalized disk IO BW needed by this POD. It then loops through each node in the candidate node list and checks this needed disk IO request of the POD against each node’s available disk IO capacity saved in the local cache to generate an updated candidate list.
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: ga_pod
+  annotations:
+     blockio.kubernetes.io/throughput: |
+{"rbps": "20M","wbps": "30M","blocksize": "4k"}
+spec:
+  containers:
+    - name: xxxServer
+      image: xxx
+       volumeMounts:  
+        - name: xxx-storage
+          mountPath: /data/xxx
+  volumes:
+  - name: xxx-storage
+     emptyDir: {}
+```
+
+#### Score Plugin
+
+During the score phase, the scheduler plugin gives a score to each node in the candidate list based on a scoring policy.
+```
+T = Node's available disk IO capacity
+R = Pod's needed disk IO bandwidth
+```
+For the Most Allocated policy:
+
+$$ score = { T - R \over T} $$
+For the Least Allocated policy:
+$$ score = {R \over T} $$
+
+#### Reserve Plugin
+
+During the reserve phase, the scheduler plugin updates the selected node’s available disk IO capacity by deducting the pod’s needed disk IO resource. In addition, it adds the pod to the ReservedPods list, tags the list with a Generation, updates the new ReservedPods list to the CR in the API server and then the IO Driver which is watching the CR would be notified of the change. The Generation, similar to the Kubernetes’s [ResourceVersion](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions), ensures that only the update on the latest generation ReservedPod list is saved to the scheduler plugin’s cache.
```

Comment:
What's the motivation to persist the in-memory reserving-pods list to CR? Ideally, I'd expect the _transient_ reserve operation (in scheduler) is transparent to any component that reads the CR, so we don't need to issue API call to update the CR in a scheduling cycle. (iow, issuing blocking API calls in a scheduling flow is an anti-pattern)

Reply:
In our scenario, the IO Driver monitors and reports real-time available disk IO to the scheduler plugin for pod scheduling decisions. The accuracy of this report is crucial and can be affected by the "ReservedPods list." In some cases, the IO Driver's "ReservedPods list" may not match the updated list in the scheduler cache during a new pod's reservation time window. This discrepancy can lead to overprovisioning on a node, as the IO Driver doesn't consider the new pod's disk IO consumption.
To address this issue, we implement a solution by tagging the "ReservedPods list" with a generation. At the reserve hook point, the scheduler plugin increments the generation, tags the new "ReservedPods list" with the incremented generation, and notifies the IO Driver of the updated "ReservedPods list" via the CR. This ensures that the IO Driver considers the new pod's IO consumption when reporting real-time available disk IO. The CR acts as a vital communication link, preventing incorrect reports of real-time disk IO and ensuring accurate scheduling decisions. Its timely update enables the scheduler's cache to reflect the most current and accurate information about available disk IO.                                                                 BTW, the API call to update the "ReservedPods list" via the CR is asynchronous and non-blocking.

Reply:
> What's the motivation to persist the in-memory reserving-pods list to CR? Ideally, I'd expect the _transient_ reserve operation (in scheduler) is transparent to any component that reads the CR, so we don't need to issue API call to update the CR in a scheduling cycle. (iow, issuing blocking API calls in a scheduling flow is an anti-pattern)

In our scenario, the IO Driver monitors and reports real-time available disk IO to the scheduler plugin for pod scheduling decisions. The accuracy of this report is crucial and can be affected by the "ReservedPods list." In some cases, the IO Driver's "ReservedPods list" may not match the updated list in the scheduler cache during a new pod's reservation time window. This discrepancy can lead to overprovisioning on a node, as the IO Driver doesn't consider the new pod's disk IO consumption.
To address this issue, we implement a solution by tagging the "ReservedPods list" with a generation. At the reserve hook point, the scheduler plugin increments the generation, tags the new "ReservedPods list" with the incremented generation, and notifies the IO Driver of the updated "ReservedPods list" via the CR. This ensures that the IO Driver considers the new pod's IO consumption when reporting real-time available disk IO. The CR acts as a vital communication link, preventing incorrect reports of real-time disk IO and ensuring accurate scheduling decisions. Its timely update enables the scheduler's cache to reflect the most current and accurate information about available disk IO.                                                                 BTW, the API call to update the "ReservedPods list" via the CR is asynchronous and non-blocking.

Reply:
> This discrepancy can lead to overprovisioning on a node, as the IO Driver doesn't consider the new pod's disk IO consumption.

Thanks for the explanation. Once you prepare for the PR, I'd like this case to be implemented as an integration test so that it can prove the "ReservedPods list" is mandatory.